### DISCUSSION TITLE
Recommended Information Extraction Method for PDF-Resumes?

### DISCUSSION THREAD

**Discussion by MichaelRinger** [explosion/spaCy#8516] [Help: Coding & Implementations] [answered]
Hello,

So for the start I thought about using just a normal BERT and only mark the entities that want to extract in my training data. But does it also makes sense to create a label for the label "english" (see example below) to get better results or to use relation extraction at this semi-form like data? Or does the behaviour of BERT recognizes automatically that after the string "english: " is going be a grade?

For a simple example at extracting grades from a resume:
input:
"english: 2"
-> do i need only need to annotate "2" as grade_english or is it recommended to do something else

output:
grade_english: "2"

wanted labels:
- firstname
- lastname
- last_job_title
- graduation
- grade_english
- grade_math
- grade_economy

**Comment by polm** [comment, marked as answer]
My understanding is that annotating pure numbers in isolation as values generally doesn't work for NER, and it's better to model this as a relation extraction problem. For something like "english: 2" you might have better luck using Matcher patterns. If you have complete sentences you should look at the DependencyMatcher; see [this example](https://spacy.io/usage/linguistic-features#:~:text=the%20dependency%20parse%20can%20be%20a%20useful%20tool%20for%20information%20extraction) in the docs.

**Reply by MichaelRinger** [reply]
Ok, but lets say i have text instead of numbers. So when i would have "Name: Jane Doe" where Jane is a firstname entity and Doe is a lastname Entity and "Name" doesnt have a label. Does a standard token classification BERT knows that after "Name" is going to be a firstname?

**Reply by polm** [reply]
So an NER model uses several things to detect entities. These include:

- token identity (the word is "Jane")
- token shape (the word is alphabetic and capitalized)
- prefixes and suffixes
- context (surrounding words)

With numbers, for the first three things you only know if it's a number or not. Maybe you can guess if it's a date if it's around 2000, say, but that's not always safe. You basically only have context. 

In contrast "Jane Doe" is basically always going to be a name regardless of context, and BERT or even simpler models will know this. But it doesn't necessarily know about "Name:" unless that was in the training data. spaCy models are trained on mostly newspaper articles, so lists of key-value pairs like that are definitely not in the training data. In general the spaCy models are trained on complete, grammatical sentences, so if you text isn't much like that they might not help you much.

If you have a bunch of lists of stuff like that, which I guess you might on resumes, NER is probably not the right approach, and this might be more of what's called an "attribute extraction" problem. I worked with information extraction like that in an ecommerce setting before, and generally what you do is extract the key-value pairs and then try to map them to your schema. If you're working from resumes you might have trouble figuring out the key-value boundaries at first

**Reply by MichaelRinger** [reply]
Thanks for the answer.
Is it also possible to determine if spacy matcher matches near to each other? That could be helpfull especially at extracting data from address blocks. For example if a match of the name-pattern is prefered when its near a match of a email-pattern. Or when want to extract skills with a PhraseMatcher only after the string "Skills, Competencies"

**Reply by polm** [reply]
When the matcher returns tokens, you can get their position in the document as usual (`token.i`). You can use that manually, or incorporate information in a match pattern, but there's not really a good way to tell the matcher that matches of some types should be close in a general way - that's more the kind of thing a statistical model would learn. 

What you can do is write rules that match with many false positives, and then write code that post-processes the matches and uses features like nearby strings to remove false positives.

**Reply by MichaelRinger** [reply]
Is it a problem when some parsed pdf resume texts, about 20%, are columnwise parsed. So the text looks something like this:
"Jane Doe Adress: Phone: Email 177 Great Portland Street, London W5W 6PQ +44 (0)20 7666 8555 christoper.morgan@gmail.com"

I am using pdfminer.six, it does seem to have mostly strict rules for parsing the layout.

**Reply by polm** [reply]
spaCy won't be able to learn the column alignments in text like that (I think even a human would have trouble until they realized what happened), but it won't break the whole model. I would recommend you try training and see how it goes.
